File: src/server/ecs/systems/removeMissingModels.lua
local ReplicatedStorage = game:GetService("ReplicatedStorage")

local Matter = require(ReplicatedStorage.Packages.matter)
local components = require(ReplicatedStorage.paradoxical.ecs.components)

--[=[
    @function removeMissingModels
    @within SystemServer

    Removes entity or model when either of them are missing.
]=]
local function removeMissingModels(world)
	for id, model in world:query(components.Model) do
		for _ in Matter.useEvent(model.model, "AncestryChanged") do
			if model.model:IsDescendantOf(game) == false then
				world:remove(id, components.Model)
				break
			end
		end
		if not model.model.PrimaryPart then
			world:remove(id, components.Model)
		end
	end

	for _id, modelRecord in world:queryChanged(components.Model) do
		if modelRecord.new == nil then
			if modelRecord.old and modelRecord.old.model then
				modelRecord.old.model:Destroy()
			end
		end
	end
end

return removeMissingModels

File: src/server/systems/removeMissingModels.ts
import { useEvent, World } from "@rbxts/matter";
import { Players } from "@rbxts/services";
import { Model } from "shared/ecs/components";
import { coreCallback } from "shared/net";

/**
 * A system that removes missing {@link Model | Models}.
 *
 * If a model is removed from the game, this system will remove the
 * corresponding model from the world.
 *
 * If a model is removed from the world, this system will remove the
 * corresponding model from the game.
 */
function removeMissingModels(world: World): void {
	for (const [id, model] of world.query(Model)) {
		if (!model.model) continue;
		for (const _ of useEvent(model.model, "AncestryChanged")) {
			if (!model.model.IsDescendantOf(game)) {
				world.remove(id, Model);
				coreCallback.resetCallback.sendTo(undefined, Players.GetPlayerByUserId(id) as Player);
				break;
			}
		}
	}

	for (const [_, record] of world.queryChanged(Model)) {
		if (record.new) continue;
		record.old?.model?.Destroy();
	}
}

export = removeMissingModels;
